### PR TITLE
bug 1680504. Create collector clusterrole to watch pods/ns

### PR DESCRIPTION
**Pull request by jcantrill**
This PR fixes https://bugzilla.redhat.com/show_bug.cgi?id=1680504 by:

* creates a clusterrole for the logcollector to be able to watch all pods and namespaces

### CODE REVIEW

**Comment by richm** [comment]
I suggest setting fluentd with loglevel trace or debug to see the metadata cache stats to see if it is reporting any problems, and the cache stats appear to be changing correctly

**Comment by jcantrill** [comment]
@ewolinetz @richm Note rsyslog e2e test failed. Temporarily disabled for 4.0 since its no longer techpreview.  We'll have to re-enable for 4.x

**Comment by richm** [comment]
looking at the full test output - https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_cluster-logging-operator/106/pull-ci-openshift-cluster-logging-operator-master-e2e-operator/155/build-log.txt - fluentd failed too - so it might not be a problem with rsyslog

File: pkg/k8shandler/collection.go [hunk 1]
```diff
@@ -83,48 +83,50 @@ func CreateOrUpdateCollection(cluster *logging.ClusterLogging) (err error) {
 		if retryErr != nil {
 			return fmt.Errorf("Failed to update Cluster Logging Fluentd status: %v", retryErr)
 		}
-	} else {
-		removeFluentd(cluster)
```
Comment (ewolinetz) [inline pkg/k8shandler/collection.go]: this needs to stay here, this is in the case where a user removes the fluentd spec from the CR after it has been initially installed

File: pkg/k8shandler/collection.go [hunk 2]
```diff
@@ -249,9 +264,8 @@ func createOrUpdateCollectorServiceAccount(cluster *logging.ClusterLogging) erro
 	subject.APIGroup = ""
 
 	collectorReaderClusterRoleBinding := utils.NewClusterRoleBinding(
-		"openshift-logging-collector-cluster-reader",
-		cluster.Namespace,
-		"cluster-reader",
+		"cluster-logging-metadata-reader",
+		clusterRole.ObjectMeta.Name,
```
Comment (ewolinetz) [inline pkg/k8shandler/collection.go]: you can shorten this to just `clusterRole.Name` :)

File: pkg/k8shandler/collection.go [hunk 3]
```diff
@@ -261,7 +275,7 @@ func createOrUpdateCollectorServiceAccount(cluster *logging.ClusterLogging) erro
 
 	err = sdk.Create(collectorReaderClusterRoleBinding)
 	if err != nil && !errors.IsAlreadyExists(err) {
-		return fmt.Errorf("Failure creating Log collector cluster-reader role binding: %v", err)
+		return fmt.Errorf("Failure creating Log collector '%s' cluster role binding: %v", collectorReaderClusterRoleBinding.ObjectMeta.Name, err)
```
Comment (ewolinetz) [inline pkg/k8shandler/collection.go]: same here, you can drop `ObjectMeta`

File: pkg/k8shandler/collection.go [hunk 4]
```diff
@@ -241,6 +243,19 @@ func createOrUpdateCollectorServiceAccount(cluster *logging.ClusterLogging) erro
...
+	// create clusterrole for logcollector to retrieve metadata
+	clusterrules := utils.NewPolicyRules(
+		utils.NewPolicyRule(
+			[]string{""},
+			[]string{"pods", "namespaces"},
+			nil,
+			[]string{"get", "list", "watch"},
+		),
+	)
+	clusterRole, err := utils.CreateClusterRole("metadata-reader", clusterrules, cluster)
```
Comment (ewolinetz) [inline pkg/k8shandler/collection.go]: Can you update this to follow the other role creation patterns for now?

File: pkg/k8shandler/collection.go (hunk 3 above)
Comment (ewolinetz) [reply pkg/k8shandler/collection.go]: if you make it `%q` it will quote the string for you with fmt

**Comment by ewolinetz** [comment]
@jcantrill you need to add the permission to `*` `clusterroles` for the clo SA in `manifests/03-role.yaml`

File: pkg/k8shandler/collection.go (hunk 4 above)
Comment (jcantrill) [reply pkg/k8shandler/collection.go]: I'm not sure what you are asking here.  I added code to a reusable method that does the same thing as if it was inlined.

File: pkg/k8shandler/collection.go (hunk 1 above)
Comment (jcantrill) [reply pkg/k8shandler/collection.go]: reverted but needs to be pushed

File: manifests/03-role.yaml [hunk 5]
```diff
@@ -100,6 +100,20 @@ rules:
 - apiGroups:
   - rbac.authorization.k8s.io
   resources:
+  - clusterroles
   - clusterrolebindings
   verbs:
   - "*"
+---
+kind: ClusterRoleBinding
```
Comment (ewolinetz) [inline manifests/03-role.yaml]: for consistency, can you move this to `manifests/04-role-binding.yaml`?

Comment (jcantrill) [reply manifests/03-role.yaml]: @ewolinetz updated
